R/data_access.R

```diff
@@ -0,0 +1,175 @@
+#' @title Get lab measurements from FinnGen data
+#' @param all_labs data frame with lab measurements
+#' @param lablist vector of lab measurement concept IDs
+#' @param require_values logical, if TRUE, only return rows with non-missing MEASUREMENT_VALUE_HARMONIZED
+#' @param return_cols vector of column names to return
+#' @param finngen_ids vector of FINNGENIDs to filter the data
+#' @param lazy logical, if TRUE, return a lazy tbl object
+#' @param covariates Optional data frame with covariate data (e.g., sex, age at death)
+#' @param covariate_cols Optional character vector of column names to add from the covariates data frame
+#' @return data frame with lab measurements
+#' @export
+#' @importFrom dplyr %>% filter select collect all_of left_join distinct mutate
+#' @import stringr
+get_lab_measurements <- function(all_labs, lablist, require_values=TRUE,
+                                 return_cols=c("FINNGENID","OMOP_CONCEPT_ID", "EVENT_AGE", "MEASUREMENT_VALUE_HARMONIZED"),
+                                 finngen_ids=NULL, lazy=FALSE,
+                                 covariates=NULL, covariate_cols=NULL) {
+
+  # Store the base columns from all_labs (without covariate columns)
+  # Covariate columns will be added later via the join
+  base_return_cols <- unique(c("OMOP_CONCEPT_ID", return_cols))
+
+  # Only select columns that exist in all_labs at this point
+  # Filter out any covariate_cols from base_return_cols if they were accidentally included
+  if (!is.null(covariate_cols)) {
+    base_return_cols <- setdiff(base_return_cols, covariate_cols)
+    # But ensure we keep the default return_cols
+    base_return_cols <- unique(c("OMOP_CONCEPT_ID", "FINNGENID", "EVENT_AGE", "MEASUREMENT_VALUE_HARMONIZED",
+                                  setdiff(return_cols, covariate_cols)))
+  }
+
+  # Ensure robust matching regardless of column/vector types (character vs numeric)
+  # The OMOP_CONCEPT_ID column may be stored as DECIMAL in the parquet file
+  # We need to cast it to character BEFORE filtering to avoid type conversion errors
+  # Convert lablist to character for consistent comparison
+  lablist_chr <- as.character(lablist)
+
+  # Cast OMOP_CONCEPT_ID to character in the database query before filtering
+  # This ensures the comparison works regardless of the storage type
+  labs <- all_labs %>%
```

Comment:
we can leave this for now but thinking that we could also enforce datatypes when loading instead of keep on casting where needed. Could add as a future development so that db config could have optional datatype schema and loading would error out if wrong data format in the parquet.